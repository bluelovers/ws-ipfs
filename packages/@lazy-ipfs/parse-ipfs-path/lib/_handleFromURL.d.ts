import { IParsePathResult } from './types';
export declare const subdomainGatewayPattern: RegExp;
export declare function _handleFromURL(input: string): IParsePathResult | string;
