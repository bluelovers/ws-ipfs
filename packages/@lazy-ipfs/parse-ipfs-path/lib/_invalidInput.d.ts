export declare function _invalidInput(input: unknown): boolean;
