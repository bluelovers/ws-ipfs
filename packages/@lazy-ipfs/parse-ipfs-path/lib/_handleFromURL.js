"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._handleFromURL = exports.subdomainGatewayPattern = void 0;
const tslib_1 = require("tslib");
const is_ipfs_1 = tslib_1.__importDefault(require("is-ipfs"));
const lazy_url_1 = require("lazy-url");
const is_cid_1 = require("@lazy-ipfs/is-cid");
exports.subdomainGatewayPattern = /^https?:\/\/([^/]+)\.(ip[fn]s)\.[^/?]+/;
function _handleFromURL(input) {
    var _a;
    input = (0, is_cid_1._url_href)(input);
    if (is_ipfs_1.default.cidPath(input)) {
        return input;
    }
    else if (!/^\w+(?:[+\-]\w+)?:\/\//.test(input)) {
        return;
    }
    let url = new lazy_url_1.LazyURL(input);
    let parts;
    if (is_ipfs_1.default.subdomain(url.origin)) {
        parts = url.origin.match(exports.subdomainGatewayPattern);
        return {
            ns: parts[2],
            hash: parts[1],
            path: url.pathname,
        };
    }
    else if (is_ipfs_1.default.cid((parts = url.host.split('.'))[0])) {
        return {
            ns: ((_a = parts[1]) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === "ipns" /* EnumParsePathResultNs.ipns */
                ? "ipns" /* EnumParsePathResultNs.ipns */
                : "ipfs" /* EnumParsePathResultNs.ipfs */,
            hash: parts[0],
            path: url.pathname,
        };
    }
    else if (is_ipfs_1.default.path(url.pathname)) {
        return url.pathname;
    }
}
exports._handleFromURL = _handleFromURL;
//# sourceMappingURL=_handleFromURL.js.map