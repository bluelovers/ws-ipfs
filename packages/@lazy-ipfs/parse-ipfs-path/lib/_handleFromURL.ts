import isIPFS from 'is-ipfs';
import { LazyURL } from 'lazy-url';
import { IParsePathResult } from './types';
import { EnumParsePathResultNs } from './asserts';
import { _url_href } from '@lazy-ipfs/is-cid/index';

export const subdomainGatewayPattern = /^https?:\/\/([^/]+)\.(ip[fn]s)\.[^/?]+/

export function _handleFromURL(input: string): IParsePathResult | string
{
	input = _url_href(input as any);

	if (isIPFS.cidPath(input))
	{
		return input
	}
	else if (!/^\w+(?:[+\-]\w+)?:\/\//.test(input))
	{
		return
	}

	let url = new LazyURL(input);
	let parts: RegExpMatchArray;

	if (isIPFS.subdomain(url.origin))
	{
		parts = url.origin.match(subdomainGatewayPattern);

		return {
			ns: parts[2] as any,
			hash: parts[1],
			path: url.pathname,
		}
	}
	else if (isIPFS.cid((parts = url.host.split('.'))[0]))
	{
		return {
			ns: parts[1]?.toLowerCase() === EnumParsePathResultNs.ipns
				? EnumParsePathResultNs.ipns
				: EnumParsePathResultNs.ipfs,
			hash: parts[0],
			path: url.pathname,
		}
	}
	else if (isIPFS.path(url.pathname))
	{
		return url.pathname
	}
}

