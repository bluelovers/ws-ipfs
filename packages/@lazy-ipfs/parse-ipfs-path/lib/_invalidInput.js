"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._invalidInput = void 0;
const util_1 = require("./util");
function _invalidInput(input) {
    let type = typeof input;
    return !(0, util_1._isStringObject)(type) && type !== 'function' && type !== 'object' || !Boolean(input);
}
exports._invalidInput = _invalidInput;
//# sourceMappingURL=_invalidInput.js.map