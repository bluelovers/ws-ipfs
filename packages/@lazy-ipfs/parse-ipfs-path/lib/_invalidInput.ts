import { _isStringObject } from './util';

export function _invalidInput(input: unknown)
{
	let type = typeof input;
	return !_isStringObject(type) && type !== 'function' && type !== 'object' || !Boolean(input)
}
