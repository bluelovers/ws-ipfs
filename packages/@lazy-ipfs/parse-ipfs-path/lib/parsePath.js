"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePath = void 0;
const tslib_1 = require("tslib");
const err_code_1 = (0, tslib_1.__importDefault)(require("err-code"));
const parsePathCore_1 = require("./parsePathCore");
function parsePath(input, options) {
    try {
        return (0, parsePathCore_1.parsePathCore)(input);
    }
    catch (e) {
        if (!(options === null || options === void 0 ? void 0 : options.noThrow) && !(options === null || options === void 0 ? void 0 : options.unsafeReturn)) {
            throw (0, err_code_1.default)(e, {
                originalInput: input,
                options,
            });
        }
    }
    if (options === null || options === void 0 ? void 0 : options.unsafeReturn) {
        return {
            ns: "ipfs" /* ipfs */,
            hash: input,
            path: '',
        };
    }
}
exports.parsePath = parsePath;
//# sourceMappingURL=parsePath.js.map