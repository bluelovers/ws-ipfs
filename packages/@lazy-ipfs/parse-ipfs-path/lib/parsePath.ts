import isValidDomain from 'is-valid-domain';
import CID from 'cids';
import { isCID, toCID } from '@lazy-ipfs/to-cid';

export const enum EnumParsePathResultNs
{
	'ipfs' = 'ipfs',
	'ipns' = 'ipns',
}

export interface IParsePathResult
{
	ns: 'ipfs' | 'ipns' | EnumParsePathResultNs;
	hash: string;
	path: string;

	toString(): string;

	toJSON(): string;
}

/**
 * @see https://github.com/tableflip/dweb-path
 */
export function parsePath(input: string | Buffer | CID)
{
	let ns: string, hash: string, path: string

	if (Buffer.isBuffer(input) || isCID(input))
	{
		hash = toCID(input).toBaseEncodedString()

		ns = EnumParsePathResultNs.ipfs
		path = ''
	}
	else if (typeof input === 'string' || Object.prototype.toString.call(input) === '[object String]')
	{
		// Ensure leading slash
		if (input[0] !== '/')
		{
			input = `/${input}`
		}

		// Remove trailing slash
		if (input[input.length - 1] === '/')
		{
			input = input.slice(0, -1)
		}

		const parts = input.split('/')

		if (parts[1] === EnumParsePathResultNs.ipfs || parts[1] === EnumParsePathResultNs.ipns)
		{
			try
			{
				hash = strToCidToStr(parts[2])
			}
			catch (err)
			{
				// If IPNS then this could be a domain name
				if (parts[1] === EnumParsePathResultNs.ipns && isValidDomain(parts[2]))
				{
					hash = parts[2]
				}
				else
				{
					throw err
				}
			}

			ns = parts[1]
			path = parts.slice(3).join('/')
		}
		else
		{
			// Is parts[1] a CID?
			try
			{
				hash = strToCidToStr(parts[1])
			}
			catch (err)
			{
				throw new Error(`Unknown namespace: ${parts[1]}`)
			}

			ns = EnumParsePathResultNs.ipfs
			path = parts.slice(2).join('/')
		}

		// Ensure leading slash on non empty path
		if (path.length)
		{
			path = `/${path}`
		}
	}
	else
	{
		throw new Error('Invalid path') // What even is this?
	}

	const toString = () => `/${ns}/${hash}${path}`

	return Object.defineProperties({} as IParsePathResult, {
		ns: { value: ns, enumerable: true },
		hash: { value: hash, enumerable: true },
		path: { value: path, enumerable: true },
		toString: { value: toString },
		toJSON: { value: toString },
	})
}

export function resultToPath(result: IParsePathResult)
{
	return `/${result.ns}/${result.hash}${result.path}`
}

export function strToCidToStr(str: string)
{
	return toCID(str).toString()
}
