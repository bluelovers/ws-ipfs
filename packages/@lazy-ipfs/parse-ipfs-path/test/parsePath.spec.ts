import { parsePath, resultToPath } from '../lib/parsePath';
import CID from 'cids';

test('should parse input as string CID', () => {
	const input = 'QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn'
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toBe('QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn')
	expect(path.path).toBe('')
	expect(resultToPath(path)).toBe('/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn')
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should parse input as buffer CID', () => {
	const input = Buffer.from('017012207252523e6591fb8fe553d67ff55a86f84044b46a3e4176e10c58fa529a4aabd5', 'hex')
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toMatchSnapshot()
	expect(path.path).toBe('')
	expect(resultToPath(path)).toBe(`/ipfs/${path.hash}`)
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should parse input as CID instance', () => {
	const input = new CID('zdpuArHMUAYi3VtD3f7iSkXxYK9xo687SoNf5stAQNCMzd77k')
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toBe('zdpuArHMUAYi3VtD3f7iSkXxYK9xo687SoNf5stAQNCMzd77k')
	expect(path.path).toBe('')
	expect(resultToPath(path)).toBe('/ipfs/zdpuArHMUAYi3VtD3f7iSkXxYK9xo687SoNf5stAQNCMzd77k')
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should parse input as string with path and without namespace', () => {
	const input = 'QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to'
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toBe('QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn')
	expect(path.path).toBe('/path/to')
	expect(resultToPath(path)).toBe('/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to')
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should parse input as string without leading slash', () => {
	const input = 'ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to'
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toBe('QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn')
	expect(path.path).toBe('/path/to')
	expect(resultToPath(path)).toBe('/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to')
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should parse input as string with trailing slash', () => {
	const input = '/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to/'
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toBe('QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn')
	expect(path.path).toBe('/path/to')
	expect(resultToPath(path)).toBe('/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to')
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should not alter CID encoding', () => {
	// base32 encoded CID
	const input = 'bafybeic7rclblnrf76dtmmrdfxix2aq7hnvms5hlb3f7zko52kvv42mb64'
	const path = parsePath(input)

	expect(path.ns).toBe('ipfs')
	expect(path.hash).toBe('bafybeic7rclblnrf76dtmmrdfxix2aq7hnvms5hlb3f7zko52kvv42mb64')
	expect(path.path).toBe('')
	expect(resultToPath(path)).toBe('/ipfs/bafybeic7rclblnrf76dtmmrdfxix2aq7hnvms5hlb3f7zko52kvv42mb64')
	expect(resultToPath(path)).toMatchSnapshot()

	expect(path).toMatchSnapshot()
})

test('should allow IPNS path', () => {
	const input = '/ipns/yourdomain.name'
	const path = parsePath(input)

	expect(path.ns).toBe('ipns')
	expect(path.hash).toBe('yourdomain.name')
	expect(path.path).toBe('')
	expect(resultToPath(path)).toBe('/ipns/yourdomain.name')
	expect(resultToPath(path)).toMatchSnapshot()
})

test('should throw on unknown namespace', () => {
	const input = '/junk/stuff'
	expect(() => parsePath(input)).toThrow()
})

test('should throw on invalid CID in string', () => {
	const input = '/ipfs/notACID/some/path'
	expect(() => parsePath(input)).toThrow()
})

test('should throw on invalid CID in buffer', () => {
	const input = Buffer.from('notaCID')
	expect(() => parsePath(input)).toThrow()
})

test('should throw on invalid path', () => {
	const input = 42
	// @ts-expect-error
	expect(() => parsePath(input)).toThrow()
})

test('should only enumerate certain properties', () => {
	const input = 'QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn'
	const path = parsePath(input)
	const enumerable = ['ns', 'path', 'hash']
	expect(Object.keys(path)).toHaveLength(enumerable.length)
	expect(Object.keys(path).every(k => enumerable.includes(k))).toBeTruthy()

	expect(Object.keys(path)).toMatchSnapshot()
	expect(resultToPath(path)).toMatchSnapshot()

})
