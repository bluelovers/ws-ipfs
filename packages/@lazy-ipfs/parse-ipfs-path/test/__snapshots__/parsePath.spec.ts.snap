// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow IPNS path 1`] = `"/ipns/yourdomain.name"`;

exports[`should not alter CID encoding 1`] = `"/ipfs/bafybeic7rclblnrf76dtmmrdfxix2aq7hnvms5hlb3f7zko52kvv42mb64"`;

exports[`should not alter CID encoding 2`] = `
Object {
  "hash": "bafybeic7rclblnrf76dtmmrdfxix2aq7hnvms5hlb3f7zko52kvv42mb64",
  "ns": "ipfs",
  "path": "",
}
`;

exports[`should only enumerate certain properties 1`] = `
Array [
  "ns",
  "hash",
  "path",
]
`;

exports[`should only enumerate certain properties 2`] = `"/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn"`;

exports[`should parse input as CID instance 1`] = `"/ipfs/zdpuArHMUAYi3VtD3f7iSkXxYK9xo687SoNf5stAQNCMzd77k"`;

exports[`should parse input as CID instance 2`] = `
Object {
  "hash": "zdpuArHMUAYi3VtD3f7iSkXxYK9xo687SoNf5stAQNCMzd77k",
  "ns": "ipfs",
  "path": "",
}
`;

exports[`should parse input as buffer CID 1`] = `"bafybeidskjjd4zmr7oh6ku6wp72vvbxyibcli2r6if3ocdcy7jjjusvl2u"`;

exports[`should parse input as buffer CID 2`] = `"/ipfs/bafybeidskjjd4zmr7oh6ku6wp72vvbxyibcli2r6if3ocdcy7jjjusvl2u"`;

exports[`should parse input as buffer CID 3`] = `
Object {
  "hash": "bafybeidskjjd4zmr7oh6ku6wp72vvbxyibcli2r6if3ocdcy7jjjusvl2u",
  "ns": "ipfs",
  "path": "",
}
`;

exports[`should parse input as string CID 1`] = `"/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn"`;

exports[`should parse input as string CID 2`] = `
Object {
  "hash": "QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn",
  "ns": "ipfs",
  "path": "",
}
`;

exports[`should parse input as string with path and without namespace 1`] = `"/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to"`;

exports[`should parse input as string with path and without namespace 2`] = `
Object {
  "hash": "QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn",
  "ns": "ipfs",
  "path": "/path/to",
}
`;

exports[`should parse input as string with trailing slash 1`] = `"/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to"`;

exports[`should parse input as string with trailing slash 2`] = `
Object {
  "hash": "QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn",
  "ns": "ipfs",
  "path": "/path/to",
}
`;

exports[`should parse input as string without leading slash 1`] = `"/ipfs/QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn/path/to"`;

exports[`should parse input as string without leading slash 2`] = `
Object {
  "hash": "QmUmaEnH1uMmvckMZbh3yShaasvELPW4ZLPWnB4entMTEn",
  "ns": "ipfs",
  "path": "/path/to",
}
`;
